flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/JobManagerMetricGroup.java

```diff
@@ -38,7 +39,7 @@
  * tasks any more
  */
 public class JobManagerMetricGroup extends ComponentMetricGroup<JobManagerMetricGroup>
-        implements LocallyCleanableResource {
+        implements LocallyCleanableResource, GloballyCleanableResource {
```

Comment:
Please expand on why this is a global resource. Is it because it the contained metrics _may_ be persisted to an external system?
I'm confused since GloballyCleanableResource says that global resources must survive a failover, but that's generally not the case for the JMJMG.

Reply:
To add to that, what are the actual semantics when implementing both the local & global variants?
Local says they are cleaned up on a locally-terminated state, global says _only_ after a globally-terminal state.
Now which is it?

Reply:
Off-topic: The GloballyCleanableResource javadocs seem incorrect as they should link to LocallyCleanableResource in Line 3.

Reply:
My intention to add the `GloballyCleanableResource` interface to the `JobManagerMetricGroup` is based on the fact that the previous implementation (i.e. 1.14-) cleaned the `JobManagerMetricGroup` in both cases, the local cleanup and the global cleanup. I wanted to keep it like that.

Reply:
This doesn't feel right. As long as this class doesn't have any artifacts that can outlive the JM, it shouldn't implement this interface.

How about doing something like this instead?

```java
    @Override
    public ResourceCleaner createGlobalResourceCleaner(
            ComponentMainThreadExecutor mainThreadExecutor) {

        return DefaultResourceCleaner.forGloballyCleanableResources(
                        mainThreadExecutor, cleanupExecutor, retryStrategy)
                .withPrioritizedCleanup(jobManagerRunnerRegistry)
                .withRegularCleanup(jobGraphWriter)
                .withRegularCleanup(blobServer)
                .withRegularCleanup(highAvailabilityServices)
                .withRegularCleanup(ofLocalResource(jobManagerMetricGroup))
                .build();
    }

    /**
     * A simple wrapper for the resources that don't have any artifacts that can outlive the {@link
     * org.apache.flink.runtime.dispatcher.Dispatcher}, but we still want to clean up their local
     * state when we terminate locally.
     *
     * @param localResource Local resource that we want to clean during a global cleanup.
     * @return Globally cleanable resource.
     */
    private static GloballyCleanableResource ofLocalResource(
            LocallyCleanableResource localResource) {
        return localResource::localCleanupAsync;
    }

Reply:
That's actually a cleaner approach I haven't thought of. That would also apply to the `JobManagerRunnerRegistry`. It makes sense because the `DispatcherResourceCleanerFactory` is more closely tight to the `Dispatcher`. I will adapt the PR accordingly and update the JavaDoc to fit.